TestResignAPIPrimaryForward is unstable

## Flaky Test

### Which jobs are failing
```
2023-08-29T02:24:03.1901281Z panic: test timed out after 15m0s
2023-08-29T02:24:03.1901409Z running tests:
2023-08-29T02:24:03.1901628Z 	TestAPIServerForwardTestSuite (10m30s)
2023-08-29T02:24:03.1901947Z 	TestAPIServerForwardTestSuite/TestResignAPIPrimaryForward (8m21s)
2023-08-29T02:24:03.1902048Z 
2023-08-29T02:24:03.1902195Z goroutine 2571455 [running]:
2023-08-29T02:24:03.1902412Z testing.(*M).startAlarm.func1()
2023-08-29T02:24:03.1902691Z 	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:2259 +0x1fc
2023-08-29T02:24:03.1902827Z created by time.goFunc
2023-08-29T02:24:03.1903074Z 	/opt/hostedtoolcache/go/1.21.0/x64/src/time/sleep.go:176 +0x45
2023-08-29T02:24:03.1903084Z 
2023-08-29T02:24:03.1903251Z goroutine 1 [chan receive, 10 minutes]:
2023-08-29T02:24:03.1903448Z testing.(*T).Run(0xc0001861a0, {0x42f441f, 0x1d}, 0x4daf520)
2023-08-29T02:24:03.1903696Z 	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:1649 +0x856
2023-08-29T02:24:03.1903856Z testing.runTests.func1(0x0?)
2023-08-29T02:24:03.1904130Z 	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:2054 +0x85
2023-08-29T02:24:03.1904302Z testing.tRunner(0xc0001861a0, 0xc000665a30)
2023-08-29T02:24:03.1904566Z 	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:1595 +0x239
2023-08-29T02:24:03.1904821Z testing.runTests(0xc0002ed400?, {0x65a6b00, 0xc, 0xc}, {0x105c733?, 0x4?, 0x6676d60?})
2023-08-29T02:24:03.1905086Z 	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:2052 +0x897
2023-08-29T02:24:03.1905226Z testing.(*M).Run(0xc0002ed400)
2023-08-29T02:24:03.1905479Z 	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:1925 +0xb58
2023-08-29T02:24:03.1905744Z go.uber.org/goleak.VerifyTestMain({0x51e3a80, 0xc0002ed400}, {0x66787e0, 0x8, 0x8})
2023-08-29T02:24:03.1906054Z 	/home/runner/go/pkg/mod/go.uber.org/goleak@v1.1.12/testmain.go:53 +0x58
2023-08-29T02:24:03.1906279Z github.com/tikv/pd/tests/integrations/mcs/tso.TestMain(...)
2023-08-29T02:24:03.1906549Z 	/home/runner/work/pd/pd/tests/integrations/mcs/tso/server_test.go:51
2023-08-29T02:24:03.1906664Z main.main()
2023-08-29T02:24:03.1906826Z 	_testmain.go:111 +0x359
2023-08-29T02:24:03.1906839Z 
2023-08-29T02:24:03.1906984Z goroutine 41 [chan receive]:
2023-08-29T02:24:03.1907228Z go.etcd.io/etcd/pkg/logutil.(*MergeLogger).outputLoop(0xc000013bd8)
2023-08-29T02:24:03.1907771Z 	/home/runner/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/pkg/logutil/merge_logger.go:173 +0xa7
2023-08-29T02:24:03.1908037Z created by go.etcd.io/etcd/pkg/logutil.NewMergeLogger in goroutine 1
2023-08-29T02:24:03.1908567Z 	/home/runner/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/pkg/logutil/merge_logger.go:91 +0x167
2023-08-29T02:24:03.1908576Z 
2023-08-29T02:24:03.1908710Z goroutine 1018 [select]:
```

### CI link
https://github.com/tikv/pd/actions/runs/6006223367/job/16292464207?pr=6996

### Reason for failure (if possible)


### Anything else

pd-3 became leader after pd-1 and pd-2 are closing.
```
2023-08-31T11:14:05.3394133Z [2023/08/31 11:04:19.833 +00:00] [INFO] [server.go:517] ["closing server"]
2023-08-31T11:14:05.3459417Z [2023/08/31 11:04:20.516 +00:00] [INFO] [server.go:517] ["closing server"]
2023-08-31T11:14:05.3470971Z [2023/08/31 11:04:21.005 +00:00] [INFO] [server.go:1656] ["campaign API service leader ok"] [campaign-leader-name=pd3]
```
pd-3 start goroutine for keyspace manager with ctx from raft cluster.
```
2023-08-31T11:14:05.3494966Z [2023/08/31 11:04:21.089 +00:00] [INFO] [tso_keyspace_group.go:167] ["start to alloc nodes to all keyspace groups"]
```
pd-3 also start to close server.
```
2023-08-31T11:14:05.3505696Z [2023/08/31 11:04:21.102 +00:00] [INFO] [server.go:517] ["closing server"]
```
At the same time, pd-1, pd-2 have closed etcd server.
```
2023-08-31T11:14:05.3456960Z [2023/08/31 11:04:20.514 +00:00] [INFO] [etcd.go:373] ["closed etcd server"] [name=pd1] [data-dir=/tmp/pd-tests1976438160] [advertise-peer-urls="[http://127.0.0.1:43251]"] [advertise-client-urls="[http://127.0.0.1:46489]"]
2023-08-31T11:14:05.3501580Z [2023/08/31 11:04:21.099 +00:00] [INFO] [etcd.go:373] ["closed etcd server"] [name=pd2] [data-dir=/tmp/pd-tests2689180332] [advertise-peer-urls="[http://127.0.0.1:42255]"] [advertise-client-urls="[http://127.0.0.1:42235]"]
```

so pd-3 met timeout when getting etcd key and raft cluster failed to create.
```
2023-08-31T11:14:05.3552810Z [2023/08/31 11:04:31.097 +00:00] [WARN] [etcdutil.go:146] ["kv gets too slow"] [request-key=/pd/7273440577555507800/rules/] [cost=10.008023847s] [error="context deadline exceeded"]
2023-08-31T11:14:05.3553254Z [2023/08/31 11:04:31.098 +00:00] [ERROR] [etcdutil.go:151] ["load from etcd meet error"] [key=/pd/7273440577555507800/rules/] [error="[PD:etcd:ErrEtcdKVGet]context deadline exceeded: context deadline exceeded"]
2023-08-31T11:14:05.3553720Z [2023/08/31 11:04:31.099 +00:00] [ERROR] [server.go:1700] ["failed to create raft cluster"] [error="[PD:etcd:ErrEtcdKVGet]context deadline exceeded: context deadline exceeded"]
```

so pd-3 will not run `stopRaftCluster` , so it also not cancel the ctx from raftcluster, which make `allocNodesToAllKeyspaceGroups` stuck.
https://github.com/tikv/pd/blob/931bf7da09615485df40511ca6415e6489284b26/server/server.go#L1699-L1703


although `server.Close` call `s.keyspaceGroupManager.Close()`, which cancel the ctx from keyspaceGroupManager, rather than the ctx from raftcluster, so it cannot stop `allocNodesToAllKeyspaceGroups`
https://github.com/tikv/pd/blob/931bf7da09615485df40511ca6415e6489284b26/pkg/keyspace/tso_keyspace_group.go#L153-L156
https://github.com/tikv/pd/blob/931bf7da09615485df40511ca6415e6489284b26/pkg/keyspace/tso_keyspace_group.go#L168-L174

--- a/pkg/keyspace/tso_keyspace_group.go
+++ b/pkg/keyspace/tso_keyspace_group.go
@@ -167,8 +167,16 @@ func (m *GroupManager) allocNodesToAllKeyspaceGroups(ctx context.Context) {
 	log.Info("start to alloc nodes to all keyspace groups")
 	for {
 		select {
+		case <-m.ctx.Done():
+			// When the group manager is closed, we should stop to alloc nodes to all keyspace groups.
+			// Note: If raftcluster is created failed but the group manager has been bootstrapped,
+			// we need to close this goroutine by m.cancel() rather than ctx.Done() from the raftcluster.
+			// because the ctx.Done() from the raftcluster will be triggered after raftcluster is created successfully.
+			log.Info("server is closed, stop to alloc nodes to all keyspace groups")
+			return
 		case <-ctx.Done():
-			log.Info("stop to alloc nodes to all keyspace groups")
+			// When the API leader is changed, we should stop to alloc nodes to all keyspace groups.
+			log.Info("the raftcluster is closed, stop to alloc nodes to all keyspace groups")
 			return
 		case <-ticker.C:
 		}

--- a/server/cluster/cluster.go
+++ b/server/cluster/cluster.go
@@ -296,13 +296,6 @@ func (c *RaftCluster) Start(s Server) error {
 		return nil
 	}
 
-	if s.IsAPIServiceMode() {
-		err = c.keyspaceGroupManager.Bootstrap(c.ctx)
-		if err != nil {
-			return err
-		}
-	}
-
 	c.ruleManager = placement.NewRuleManager(c.storage, c, c.GetOpts())
 	if c.opt.IsPlacementRulesEnabled() {
 		err = c.ruleManager.Initialize(c.opt.GetMaxReplicas(), c.opt.GetLocationLabels())
@@ -325,6 +318,15 @@ func (c *RaftCluster) Start(s Server) error {
 	c.externalTS, err = c.storage.LoadExternalTS()
 	if err != nil {
 		log.Error("load external timestamp meets error", zap.Error(err))
+	}
+
+	// bootstrap keyspace group manager after starting other parts successfully.
+	// This order avoids a stuck goroutine in keyspaceGroupManager when it fails to create raftcluster.
+	if s.IsAPIServiceMode() {
+		err = c.keyspaceGroupManager.Bootstrap(c.ctx)
+		if err != nil {
+			return err
+		}
 	}
 
 	c.wg.Add(10)

